File: src/ReplicatedStorage/Modules/BaseModule/Relative/State/AccelerationState.lua
--!strict

local Type = require(script.Parent.Parent.Parent.Parent.Type)
local Constructor = require(script.Parent.Parent.Parent.Parent.Constructor)
local State = require(script.Parent)
local Vector3D = require(script.Parent.Parent.Parent.Parent.Parent.Modules.Libraries.Vector3D)

-- Internal type
type AccelerationState = Type.AccelerationStateEXTENSIBLE<AccelerationState,
		Type.StateEXTENSIBLE<AccelerationState,
			Type.RelativeEXTENSIBLE<AccelerationState,
				Type.BaseModuleEXTENSIBLE<AccelerationState
	>>>>
	& Constructor.AccelerationStateEXTENSIBLE<AccelerationState>
	& {
	acceleration: Type.Vector3D,
	delta: number
}

local AccelerationState: AccelerationState = { __type = "AccelerationState" :: "AccelerationState" } :: any
local AccelerationStateMT = {}

function AccelerationStateMT.__add(self: AccelerationState, other: AccelerationState): AccelerationState
	self:assertTypesMatch(other)
	assert(self:sameRelativeTree(other), "AccelerationState __add() Operands do not have the same relativeTree")
	-- local newSelf: AccelerationState, newOther: AccelerationState = self:synchronize(other)
	-- return AccelerationState.new(newSelf.acceleration + newOther:getAccelerationVector(newSelf.delta), newSelf.delta)
	if self.delta ~= other.delta then
		return AccelerationState.new(self:getAccelerationVector(1) + other:getAccelerationVector(1))
	else
		return AccelerationState.new(self.acceleration + other.acceleration)
	end
end

function AccelerationStateMT.__sub(self: AccelerationState, other: AccelerationState): AccelerationState
	self:assertTypesMatch(other)
	assert(self:sameRelativeTree(other), "AccelerationState __sub() Operands do not have the same relativeTree")
	-- local newSelf: AccelerationState, newOther: AccelerationState = self:synchronize(other)
	-- return AccelerationState.new(newSelf.acceleration - newOther:getAccelerationVector(newSelf.delta), newSelf.delta)
	if self.delta ~= other.delta then
		return AccelerationState.new(self:getAccelerationVector(1) - other:getAccelerationVector(1))
	else
		return AccelerationState.new(self.acceleration - other.acceleration)
	end
end

function AccelerationStateMT.__eq(self: AccelerationState?, other: AccelerationState?): boolean
	if (self ~= nil and other ~= nil) then
		self:assertTypesMatch(other)

		if (self.acceleration == other.acceleration and self.delta == other.delta) then
			if (self:hasRelative() and other:hasRelative()) then
				return AccelerationStateMT.__eq(self:getRelative(), other:getRelative())
			else
				return self:hasRelative() == other:hasRelative()
			end
		else
			return false
		end
	else
		return self == nil and other == nil
	end
end

function AccelerationStateMT.__lt(self: AccelerationState, other: AccelerationState): boolean
	self:assertTypesMatch(other)
	return self:getAbsoluteAcceleration() < other:getAbsoluteAcceleration()
end

function AccelerationStateMT.__le(self: AccelerationState, other: AccelerationState): boolean
	self:assertTypesMatch(other)
	return self:getAbsoluteAcceleration() <= other:getAbsoluteAcceleration()
end

-- Constructors

--[=[
	Creates a new AccelerationState instance.
]=]
function AccelerationState.new(acceleration: Type.Vector3D, delta: number?, relativeTo: AccelerationState?): AccelerationState
	local self: AccelerationState = table.clone(AccelerationState) :: any
	self.acceleration = acceleration
	self.delta = delta or 1

	local metatable = table.clone(AccelerationStateMT)
	metatable.__index = State.new(relativeTo)

	setmetatable(self, metatable)

	return self
end

--[=[
	Clones an AccelerationState instance, altering its delta.
]=]
function AccelerationState.newWithDelta(accelerationState: AccelerationState, delta: number?): AccelerationState
	local self: AccelerationState = table.clone(AccelerationState) :: any
	local newDelta: number = delta or 1
	self.acceleration = accelerationState:getAccelerationVector(newDelta)
	self.delta = newDelta

	local metatable = table.clone(AccelerationStateMT)
	if accelerationState:hasRelative() then
		metatable.__index = State.new(accelerationState:getRelative())
	else
		metatable.__index = State.new()
	end

	setmetatable(self, metatable)

	return self
end

-- Methods

--[=[
	Returns the acceleration vector (velocity over delta).
]=]
function AccelerationState.getAccelerationVector(self: AccelerationState, delta: number?): Type.Vector3D
	if delta == nil then
		return self.acceleration
	else
		return self.acceleration * (delta / self.delta)
	end
end

function AccelerationState.getAbsoluteAcceleration(self: AccelerationState, delta: number?): Type.Vector3D
	if self:hasRelative() then
		return self:getAccelerationVector(delta) + self:getRelative():getAbsoluteAcceleration(delta)
	else
		return self:getAccelerationVector(delta)
	end
end

function AccelerationState:getDelta(): number
	return self.delta
end

--[=[
	Returns a new AccelerationState relative to the current relativeTo's relativeTo.

	<code>self: &nbsp;&nbsp;a-b-c-d-self</code><br>
	<code>result: a-b-c-result</code>

	@return The resultant AccelerationState.
]=]
function AccelerationState:consolidateOnce(delta: number?): AccelerationState
	assert(self:hasRelative(), "consolidateOnce() cannot be called on a AccelerationState with no RelativeTo")
	local relativeTo: AccelerationState = self:getRelative()

	return AccelerationState.new(
		self:getAccelerationVector(delta) + relativeTo:getAccelerationVector(delta),
		delta,
		relativeTo:getRelativeOrNil()
	)
end

--[=[
	Synchronizes this AccelerationState with another such that they have the same RelativeTo.

	<code>self: &nbsp;&nbsp;a-b-c-d-self</code><br>
	<code>other: &nbsp;a-b-e-other</code><br>
	<code>result: a-b-selfResult, a-b-otherResult</code>

	@param other The other AccelerationState to synchronize with.
	@return The synchronized AccelerationStates as a tuple of self and other, in that order.
]=]
function AccelerationState:synchronize(other: AccelerationState): (AccelerationState, AccelerationState)
	local convergenceItem: AccelerationState = self:convergenceItem(other) or AccelerationState.new(Vector3D.zero)

	local selfTree: { AccelerationState } = self:getRelativeTree()
	local selfTrimmedAcceleration: Type.Vector3D = Vector3D.zero

	for i = 1, self:convergenceIndex(other) - 1 do
		selfTrimmedAcceleration += selfTree[i]:getAccelerationVector(convergenceItem.delta)
	end

	local otherTree: { AccelerationState } = other:getRelativeTree()
	local otherTrimmedAcceleration: Type.Vector3D = Vector3D.zero

	for i = 1, other:convergenceIndex(self) - 1 do
		otherTrimmedAcceleration += otherTree[i]:getAccelerationVector(convergenceItem.delta)
	end

	local selfResult: AccelerationState = AccelerationState.new(
			convergenceItem.acceleration + selfTrimmedAcceleration,
			convergenceItem.delta
		)
	local otherResult: AccelerationState = AccelerationState.new(
			convergenceItem.acceleration + otherTrimmedAcceleration,
			convergenceItem.delta
		)

	assert(
		selfResult:getRelativeOrNil() == otherResult:getRelativeOrNil()
		and self:getAbsoluteAcceleration() == selfResult:getAbsoluteAcceleration()
		and other:getAbsoluteAcceleration() == otherResult:getAbsoluteAcceleration(),
		"something wrong in the calcs!"
	)

	return selfResult, otherResult
end

--[=[
	Matches the RelativeTo tree of other with this AccelerationState.
	The result's delta will also match that of self.

	<code>self: &nbsp;&nbsp;a-b-c-d-self</code><br>
	<code>other: &nbsp;a-b-e-other</code><br>
	<code>result: a-b-c-d-otherResult</code>

	@param other The other AccelerationState to match with.
	@return The synchronized other AccelerationState.
]=]
function AccelerationState:matchRelative(other: AccelerationState): AccelerationState
	local convergenceIndex: number = other:convergenceIndex(self)

	-- consolidate other to match with self's RelativeTo tree, and track trimmed relativeTime
	local otherIterator: AccelerationState = other
	local trimmedAcceleration: Type.Vector3D = Vector3D.zero

	for i = 1, convergenceIndex - 1 do
		trimmedAcceleration += otherIterator:getAccelerationVector(self.delta)
		otherIterator = otherIterator:getRelative()
	end

	-- subtract the time exclusively between other and self, and add the excess to the newly matched result
	local selfRelativeTree: { AccelerationState } = self:getRelativeTree()
	local resultAccelerationLeftover: Type.Vector3D = trimmedAcceleration

	for i = self:convergenceIndex(other) - 1, 2, -1 do
		resultAccelerationLeftover -= selfRelativeTree[i]:getAccelerationVector(self.delta)
	end

	local result: AccelerationState = AccelerationState.new(resultAccelerationLeftover, self.delta, self:getRelativeOrNil())

	assert(
		self:getRelativeOrNil() == result:getRelativeOrNil()
		and other:getAbsoluteAcceleration(self.delta) == result:getAbsoluteAcceleration(self.delta),
		"something wrong in the calcs!"
	)

	return result
end

return (AccelerationState :: any) :: Constructor.AccelerationState

File: src/shared/Modules/BaseModule/Relative/State/AccelerationState.ts
// import { $assert } from "rbxts-transform-debug";
import Vector3D from "shared/Modules/Libraries/Vector3D";

import State from ".";

export default class AccelerationState extends State {
	public readonly acceleration: Vector3D;
	public readonly delta: number;

    // Constructors

	/**
	 * Creates a new AccelerationState instance.
	 * @param acceleration The velocity change.
	 * @param delta The time over which the velocity change is applied. Defaults to 1.
	 * @param relativeTo 
	 */
	public constructor(acceleration: Vector3D, delta?: number, relativeTo?: AccelerationState)

	/**
	 * Returns a modified version of accelerationState, with an altered delta.
	 * @param accelerationState The AccelerationState to alter.
	 * @param delta The delta used to alter accelerationState. Defaults to 1.
	 */
	public constructor(accelerationState: AccelerationState, delta?: number)

	public constructor(arg1: Vector3D | AccelerationState, arg2: number = 1, arg3?: AccelerationState) {
		let accelerationVector: Vector3D;
		let delta: number;
		let relativeTo: AccelerationState | undefined;

		if (arg1 instanceof Vector3D) { // Constructor 1
			accelerationVector = arg1;
			delta = arg2;
			relativeTo = arg3;
		} else { // Constructor 2
			const newAccelerationVector = arg1.getAccelerationVector(arg2);
			accelerationVector = newAccelerationVector;
			delta = arg2
			relativeTo = arg1.getRelativeOrUndefined();
		}

	    super(relativeTo);
		this.acceleration = accelerationVector;
		this.delta = delta;
	}

    // Arithmetic

	/**
	 * Adds two AccelerationStates.
	 * @param other The AccelerationState to add with the current instance.
	 * @returns A new AccelerationState with the delta of the current instance.
	 */
	public add(other: AccelerationState): AccelerationState {
		assert(this.sameRelativeTree(other), "AccelerationState add() operands do not have the same relativeTree");

		if (this.delta !== other.delta) {
			return new AccelerationState(
				this.getAccelerationVector(this.delta).add(other.getAccelerationVector(this.delta))
			);
		} else {
			return new AccelerationState(
				this.acceleration.add(other.acceleration),
				this.delta,
				this.getRelativeOrUndefined()
			);
		}
	}

	/**
	 * Subtracts two AccelerationStates.
	 * @param other The AccelerationState to subtract to the current instance.
	 * @returns A new AccelerationState with the delta of the current instance.
	 */
	public sub(other: AccelerationState): AccelerationState {
		assert(this.sameRelativeTree(other), "AccelerationState sub() operands do not have the same relativeTree");

		if (this.delta !== other.delta) {
			return new AccelerationState(
				this.getAccelerationVector(this.delta).sub(other.getAccelerationVector(this.delta)),
			);
		} else {
			return new AccelerationState(this.acceleration.sub(other.acceleration));
		}
	}

    // Comparisons

	override equals(other?: AccelerationState): other is AccelerationState {
		return super.equals(other)
			&& this.acceleration.equals(other.acceleration)
			&& this.delta === other.delta;
	}

	public lessThan(other: AccelerationState): boolean {
		return this.getAbsoluteAcceleration().magnitude() < other.getAbsoluteAcceleration().magnitude();
	}

	public lessOrEqual(other: AccelerationState): boolean {
		return this.getAbsoluteAcceleration().magnitude() <= other.getAbsoluteAcceleration().magnitude();
	}

	// Methods

	/**
	 * Changes the delta and change in velocity
	 * while maintaining the overall acceleration force.
	 * @param delta 
	 * @returns A new AccelerationState.
	 */
	public getAccelerationVector(delta?: number): Vector3D {
		if (delta === undefined) {
			return this.acceleration;
		} else {
			return this.acceleration.mul(delta / this.delta);
		}
	}

	public getAbsoluteAcceleration(delta?: number): Vector3D {
		if (this.hasRelative()) {
			return this.getAccelerationVector(delta).add(this.getRelative().getAbsoluteAcceleration(delta));
		} else {
			return this.getAccelerationVector(delta);
		}
	}

    override getAbsolute(): AccelerationState {
        return new AccelerationState(this.getAbsoluteAcceleration(1));
    }

	override consolidateOnce(delta?: number): AccelerationState {
		assert(this.hasRelative(), "consolidateOnce() cannot be called on a AccelerationState with no RelativeTo");
		const relativeTo = this.getRelative();

		return new AccelerationState(
			this.getAccelerationVector(delta).add(relativeTo.getAccelerationVector(delta)),
			delta,
			relativeTo.getRelativeOrUndefined(),
		);
	}

	override synchronize(other: AccelerationState): [AccelerationState, AccelerationState] {
		const convergenceItem = this.convergenceItem(other) ?? new AccelerationState(new Vector3D(0, 0, 0));

		const selfTree = this.getRelativeTree();
		let selfTrimmedAcceleration = new Vector3D(0, 0, 0);

		for (let i = 0; i < this.convergenceIndex(other) - 1; i++) {
			selfTrimmedAcceleration = selfTrimmedAcceleration.add(
				selfTree[i].getAccelerationVector(convergenceItem.delta),
			);
		}

		const otherTree = other.getRelativeTree();
		let otherTrimmedAcceleration = new Vector3D(0, 0, 0);

		for (let i = 0; i < other.convergenceIndex(this) - 1; i++) {
			otherTrimmedAcceleration = otherTrimmedAcceleration.add(
				otherTree[i].getAccelerationVector(convergenceItem.delta),
			);
		}

		const selfResult = new AccelerationState(
			convergenceItem.acceleration.add(selfTrimmedAcceleration),
			convergenceItem.delta,
		);
		const otherResult = new AccelerationState(
			convergenceItem.acceleration.add(otherTrimmedAcceleration),
			convergenceItem.delta,
		);

		assert(
			selfResult.getRelativeOrUndefined() === otherResult.getRelativeOrUndefined() &&
				this.getAbsoluteAcceleration().equals(selfResult.getAbsoluteAcceleration()) &&
				other.getAbsoluteAcceleration().equals(otherResult.getAbsoluteAcceleration()),
			"something wrong in the calcs!",
		);

		return [selfResult, otherResult];
	}

	override matchRelative(other: AccelerationState): AccelerationState {
		const convergenceIndex = other.convergenceIndex(this);

		let otherIterator = other;
		let trimmedAcceleration = new Vector3D(0, 0, 0);

		for (let i = 0; i < convergenceIndex - 1; i++) {
			trimmedAcceleration = trimmedAcceleration.add(otherIterator.getAccelerationVector(this.delta));
			otherIterator = otherIterator.getRelative();
		}

		const selfRelativeTree = this.getRelativeTree();
		let resultAccelerationLeftover = trimmedAcceleration;

		for (let i = this.convergenceIndex(other) - 2; i >= 1; i--) {
			resultAccelerationLeftover = resultAccelerationLeftover.sub(
				selfRelativeTree[i].getAccelerationVector(this.delta),
			);
		}

		const result = new AccelerationState(resultAccelerationLeftover, this.delta, this.getRelativeOrUndefined());

		assert(
			this.getRelativeOrUndefined() === result.getRelativeOrUndefined() &&
				other.getAbsoluteAcceleration(this.delta).equals(result.getAbsoluteAcceleration(this.delta)),
			"something wrong in the calcs!",
		);

		return result;
	}

	// Wrap super methods with current type

    override convergenceIndex(other: AccelerationState): number {
        return super.convergenceIndex(other);
    }

	override getRelative(): AccelerationState {
        return super.getRelative() as AccelerationState;
    }

	override getRelativeOrUndefined() : AccelerationState | undefined {
        return super.getRelativeOrUndefined() as AccelerationState | undefined;
    }

    override getRelativeTree(): AccelerationState[] {
        return super.getRelativeTree() as unknown as AccelerationState[];
    }

	override convergenceItem(other: AccelerationState): AccelerationState | undefined {
        return super.convergenceItem(other) as AccelerationState | undefined;
    }

	override deepClone(): AccelerationState {
		return this;
	}
}
